### ISSUE TITLE
KITTI Training Performance on 3dop split

### ISSUE THREAD

**Issue by jan-haug** [jwchoi384/Gaussian_YOLOv3#69]
Hi, thank you for sharing your code/settings.

I have been trying to train your model on the semi-official 3dop split for comparison, but have not managed to get reasonable and comparable AP scores.
I ran several trainings, some of which didn't converge (nan loss), which led me to reduce the learning rate as proposed in https://github.com/jwchoi384/Gaussian_YOLOv3/issues/16#issuecomment-549228854
Then, after 60k iterations the results for the car class are:
(easy/moderate/hard)
81.37, 65.17, 56.59

Do you have any suggestions on how to improve the results?
Do you think this is just due to random initialization (despite using the pretrained darknet53.conv.74 weights)?

Thank you in advance,
Jan

**Comment by jwchoi384** [comment]
Hi @insite-Jan 

Are you saying that you tried to apply Gaussian YOLOv3 to 3D object detection?
I didn't take 3D object detection into account, but I don't think random initialization will significantly degrade the accuracy.
BTW, if you want to use this repo to 3D object detection, you need to modify the output layer and localization loss function for training.

**Comment by jan-haug** [comment]
Thanks for replying so quickly!

I am still doing only 2D detection, just using a different split (the semi-official one from the [3D Object Proposals Paper](https://papers.nips.cc/paper/2015/file/6da37dd3139aa4d9aa55b8d237ec5d4a-Paper.pdf) )

Otherwise, I made no changes and also trained on 4 GPUs with 32GB each.

**Comment by jwchoi384** [comment]
@insite-Jan 

Aha, I see.
They also split the KITTI training set into train and validation sets (each containing half of the images).
When splitting, they said that their training and validation set do not come from the same video sequences.
I'm not sure exactly why the accuracy is poor, but we didn't consider whether it was the same video sequence or not when splitting. 
I think this may cause a accuracy difference.

Have you ever tried our split data for your experiments?
We provided our KITTI split data in this repo: https://github.com/jwchoi384/Gaussian_YOLOv3/issues/45#issuecomment-566437651

**Comment by jan-haug** [comment]
Yes, the choice of split causes a difference since the distributions of the training vs validation data become more different when sequences are separated.

I trained it on your split, also with original settings, on 1 GPU, and got these results:
50k steps: 88.99, 79.07, 70.56
60k steps: 89.18, 78.53, 70.02
For moderate and hard, there's still quite a large difference - are there some other settings that could make sense to try out?

Also, did you make the weights to a model trained on KITTI available by any chance?

**Comment by jwchoi384** [comment]
Hi @insite-Jan, 

When I did experiments, I used 4 GPUs for training.
BTW, you must set `batch=64` and `subdivisions=16` in `Gaussian_yolov3_KITTI_3cls.cfg` for training.
Please see the lines below:
https://github.com/jwchoi384/Gaussian_YOLOv3/blob/4cf7b926144063fe8cfdfc1b5b8c01c741dc16dc/cfg/Gaussian_yolov3_KITTI_3cls.cfg#L5
As you said you use only one GPU for training, it seems that you set the low batch and subdivisions size.

Unfortunately, I don't have the KITTI weight now, but other user got similar results using our repo and split files.
https://github.com/jwchoi384/Gaussian_YOLOv3/issues/45#issuecomment-568403179
So, I think you can get similar results to our paper if you follow our settings.

**Comment by jan-haug** [comment]
I didn't change the batchsize or subdivisions and tried (the same settings) with both 1 GPU and 4 GPUs.
The data for the 3dop split is from a 4 GPU run, because all the trainings with one GPU didn't converge (nan loss) with the original settings. On your split, this didn't happen but other than that, I didn't find any significant performance differences between training with 4 GPUs compared to one, except for the training speed of course.

Just for clarification, subdivisions determines how many steps each batch is separated into (where subdivisions=1 would mean computing the entire batch at once), not the size of the subdivisions, right?

Thank you again for going through this effort with me!

**Comment by jwchoi384** [comment]
What I mean is that the batch and subdivision size of `Gaussian_yolov3_KITTI_3cls.cfg` in the current repo are 1, so you have to change (only training step) these to `batch=64` and `subdivisions=16` for reproduce.

If you didn't change the batch and subdivisions in the cfg, you could get a lower mAP compared to ours.

I don't understand your question well.
You can see the role of batch and subdivision below:
https://github.com/AlexeyAB/darknet/issues/1736#issuecomment-427991728
If there is a subdivision, after calculating in mini-batch units (`batch/subdivision`), the model is updated in batch size units.

**Comment by jan-haug** [comment]
Ah I see. What I meant is that I didn't use a different batchsize than you did. I saw that it was 1/1 for evaluation and set it to 64/16 as you specified. A batchsize of 1 would definitely hurt training, but that is not the issue in this case.

Thanks for the explanation, that answers my question! This also means that the subdivision setting shouldn't affect final performance, but only trades off training time and memory, I assume

**Comment by jwchoi384** [comment]
No problem. In general, the larger the batch size, the better the performance.

**Comment by jan-haug** [comment]
Quick follow-up, the reason for the performance discrepancy was an issue with cudnn 8. Without that, cudnn doesn't work and performance is ~10% worse across the board.
The fix is described in https://forums.developer.nvidia.com/t/darknet-compile-error-with-cudnn-8/141115/12 (specifically the first part of that answer) in case anyone else has this issue. Basically adding another case for cudnn_version >= 8 into convolutional_layer.c and some minor Makefile modifications.

Anyways, I now get a Car AP of 90.80, 90.04, 81.01 for the 512px version on your split, which is in line with your numbers.

**Comment by jwchoi384** [comment]
Nice!
There seems to be a big difference in performance depending on cuDNN version.
As I mentioned in our paper and `Readme` file in this repo, we used cuDNNv7 in the experiments.
Happy to hear that you can obtain similar accuracy 👍.
